File: linkedlist/24.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
    var dummy, pre, cur, next *ListNode
    dummy = &ListNode{0, nil}
    dummy.Next = head
    pre = dummy
    for head != nil && head.Next != nil {
        cur = head
        next = head.Next
        cur.Next = head.Next.Next
        pre.Next = next
        next.Next = cur
        head = cur.Next
        pre = pre.Next.Next
    }
    return dummy.Next
}

File: linkedlist/24.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode *dummy = new ListNode{}, *pre, *nxt;
        dummy->next = head;
        pre = dummy;
        while (head && head->next) {
            nxt = head->next;
            pre->next = nxt;
            head->next = nxt->next;
            nxt->next = head;
            head = head->next;
            pre = pre->next->next;
        }
        return dummy->next;
    }
};
